src/pessimisticPatternMatcherUtils.cc

```diff
@@ -32,16 +32,134 @@ namespace lsst {
+            break;
+        }
+        /// Find the reference pairs that include our candidate pattern center
+        /// and sort them in increasing delta. Check this first so we don't
+        /// compute anything else if no candidates exist.
+        std::vector<size_t> candidate_range =
+                find_candidate_reference_pair_range(src_dist_array[src_idx], ref_dist_array, max_dist_rad);
+        if (candidate_range[0] == candidate_range[2]) {
+            n_fail++;
+            continue;
+        }
+
+        /// Given our length tolerance we can use it to compute a tolerance
+        /// on the angle between our spoke.
+        double src_sin_tol = max_dist_rad / (src_dist_array[src_idx] + max_dist_rad);
+
+        /// Test if the small angle approximation will still hold. This is
+        /// defined as when sin(theta) ~= theta to within 0.1% of each
+        /// other. If the implied opening angle is too large we set it to
+        /// the 0.1% threshold.
+        if (src_sin_tol > max_sin_tol) {
+            src_sin_tol = max_sin_tol;
+        }
+
+        /// Plane project the candidate source spoke and compute the cosine
+        /// and sine of the opening angle.
+        double proj_src_delta_dot =
+                ndarray::asEigenMatrix(src_delta_array[src_idx]).dot(ndarray::asEigenMatrix(src_ctr));
+
+        ndarray::Array<double, 1, 1> proj_src_delta =
+                copy(src_delta_array[src_idx] - proj_src_delta_dot * src_ctr);
+        double geom_dist_src =
+                sqrt(ndarray::asEigenMatrix(proj_src_delta).dot(ndarray::asEigenMatrix(proj_src_delta)) *
+                     proj_src_ctr_dist_sq);
+
+        /// Compute cosine and sine of the delta vector opening angle.
+        double cos_theta_src =
+                ndarray::asEigenMatrix(proj_src_delta).dot(ndarray::asEigenMatrix(proj_src_ctr_delta)) /
+                geom_dist_src;
+        Eigen::Vector3d cross_src = ndarray::asEigenMatrix(proj_src_delta)
+                                            .head<3>()
+                                            .cross(ndarray::asEigenMatrix(proj_src_ctr_delta).head<3>()) /
+                                    geom_dist_src;
+        double sin_theta_src = cross_src.dot(ndarray::asEigenMatrix(src_ctr));
+
+        /// Test the spokes and return the id of the reference object.
+        /// Return -1 if no match is found.
+        int ref_id =
+                check_spoke(cos_theta_src, sin_theta_src, ref_ctr, proj_ref_ctr_delta, proj_ref_ctr_dist_sq,
+                            candidate_range, ref_id_array, reference_array, src_sin_tol);
+        if (ref_id < 0) {
+            n_fail++;
+            continue;
+        }
+
+        /// Append the successful indices to our list. The src_idx needs
+        /// an extra iteration to skip the first and second source objects.
+        std::vector<size_t> tmpIds{ref_id, src_idx + 1};
+        output_spokes.push_back(tmpIds);
```

Comment:
You can avoid the need for a temporary variable by using `emplace`:
```suggestion
        output_spokes.emplace_back({ref_id, src_idx + 1});
```

Reply:
If there will always be two elements, consider making the elements of `output_spokes` [pairs](https://en.cppreference.com/w/cpp/utility/pair) (essentially, specialized tuples) rather than vectors.

Reply:
So the change here would be to make the return a vector of pairs? so `std::vector<std::pair<size_t, size_t>>`?

Reply:
Yes. IIRC it would get mapped to a list of tuples in Python.